### ISSUE TITLE
Stateful Service not working with 50+ partitions

### ISSUE THREAD

**Issue by francescocristallo** [Azure/service-fabric-issues#915]
Servicer Fabric live server,  version 6.1.467.9494 on Windows Server 2016
Asp.net Core 2.0 Stateful service, Remote V2

Low Key | -7968993439842526272
High Key | -7932099951695107169
500 partitions
TargetReplicaSetSize" Value="3"
MinReplicaSetSize" Value="2"

The same exact service **with 50 partitions works perfectly**.

This is what I see in SF Explorer right after the upload with no live traffic.

<img width="1849" alt="capture" src="https://user-images.githubusercontent.com/9793825/37319763-674738d4-263e-11e8-9f95-786b6ec95557.PNG">

Is there a max number of partitions I can create? If I lower it to 50 it works, but as soon as deploy a new service with more than 50 it simply doesn't work.

I tried with 80, 100, 200, 250 300, 500, only 50 or less works.
I need so many partitions for a reason (process thousands concurrent queue per second).

**Comment by guibirow** [comment]
it is likely that the SF is trying to put a secondary replica in a node that already contains a primary, and it is 'forbidden', 
You might have placement constraints that block your service being initiated in a different node(worker==true) or other nodes are rejecting the service type you are trying to deploy(as suggest in the message).
Did you try increasing the number of nodes or check if there are nodes down\deactivated?

**Comment by francescocristallo** [comment]
The nodes are all up, the Cluster was brand new. There is a placement constraint to place the Stateful services on the Node type "Worker" that have Silver reliability and 5 machines.
With 50 or less works, as soon as I increase the number of partitions goes down completely with quorum loss on all partitions.
Seems like a bug or limitation to me. 
Maybe for more than 50 partitions I need more than 5 machines?

**Comment by francescocristallo** [comment]
So I tried 100 partitions on 10 VM and it worked. As soon as I deploy a Stateful Service with more than 100 on a 10VM or 50 on 5 VM the entire service is in quorum loss.
Any idea why this is happening? Is this a known limitation of Service Fabric?

**Comment by xunlu** [comment]
There is no such limitation on service fabric side.  You should try to first run N instances of your service on a single machine to see how many you can run.  It could be running out of memory for example or run into some other system resource limitation from OS.

**Comment by francescocristallo** [comment]
The Stateful services are "empty", there is no code running apart the standard infinite while cycle from the auto-created Visual Studio template.
I just created a couple of Stateful services, set the partition numbers (I started at 500), and deployed it live with Visual Studio, and had the results above.

There should be a clear error somewhere explaining why this happens, the errors I see are just generic. This worries me a lot, I feel the orchestration system is very fragile as now.

**Comment by masnider** [comment]
@francescocristallo it's not fragile and the error messages are at least trying to tell you what's happening, we just need to put together the pieces. From what you're showing, all we can see is that the replicas aren't getting placed for some reason. Looks like a combination of things. 

TLDR: probably heap space, try EnableActivateNoWindow = true in your cluster settings (instructions below) and see if that helps, if not then you'll need that + to tweak Window's heap sizes. 

I know there's a lot going on, so let's step through some things to see what's probably happening:

First thing: placement constraints
- your Nodes have TypeName of "Workers" and "Engine". 
- Your service has a placement constraint of "worker == true"
- So: Unless you have specified additional [node properties](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-resource-manager-cluster-description#node-properties-and-placement-constraints), that placement constraint won't match any nodes. You'd need to type "NodeType == Workers" (or match it up with a custom property) to constrain it to the worker nodes. It looks like that may actually be working, but it'd be good to share us the node definition anyway so we can prove it. 
- We can actually see that the constraint is eliminating a bunch of nodes in the health report in your screenshot. So it's at least partially working.
- You should be able to check to see if your node types have any other properties and what they are by clicking on the Cluster node and then the Manifest tab, and finding the definition of your node types. Feel free to share that as it will help us understand what you're seeing.

Second thing: Service Type Disabled
- ServiceTypeDisabled is also interesting since it means that SF was successfully able to place some service, but it's failing. When a service fails on a machine a bunch of times, we [blocklist](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-resource-manager-management-integration#blocklisting-nodes) it temporarily. You can usually see what's failing by drilling into the health messages at that particular node. I don't know if your service is hitting it, because for the most part it doesn't look like we're placing your services at all because of the placement constraints not matching, but it's something to look into. If you do have custom other constraints then you are getting placed but then things are failing. Which leads us to:

Third thing: Issues with Memory resources, of some kind:
- **The error you reported is key**. That maps to 0xc0000142, which is STATUS_DLL_INIT_FAILED. A common cause of _that_ is either running out of memory or running out of heap space so that process activation is failing. (And process activation failing leads to Service Types getting disabled, so I think we're on to something :) )
- So now let's talk capacity: You're creating (a couple) of service with 500 partitions and 3 replicas a piece. So let's say 1000 partitions, which means 3000 replicas. 
- If the placement constraints are right, I **think** you have 18 nodes where you want this service created.  That means ~170 replicas per machine. That's not terribly high, but it's going to be highly dependent on the next couple of things: VM Size and [ServicePackageActivationMode](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-hosting-model#choosing-between-shared-process-and-exclusive-process-model). I can't see either of those from what you've shared with us, so if you could grab that info, that'd be great.
- So why would it work with 50 and not with 500? Well - the first thing to check is always RAM. We've see this before. Let's walk through it.
- Let's say you've got more common setting on ExclusiveProcess for ServicePackageActivationMode. That means that for each replica, you're creating a new process, loading the CLR, etc. That means 170x(workingSet) of that service replica in terms of memory. 
- Let's say that the working set of the service is 60Mb. That's pretty generous for .NET Core apps since they tend to carry around a lot of stuff with them, but we'll start there.
- So at a minimum, just to start your replicas, you'd need about 10Gb of RAM available on those VMs. Do you have that?
- Next though, let's say that you're using the _default_ ServicePackageActivationMode of SharedProcess. This means that all those replicas are getting loaded into the same process. That means they share some dependencies, but now that process will still need around 10Gb of RAM (probably a little less, but it's hard to guess). 
- Let's say though that there's plenty of RAM. Now we have to check out heap space. What's heap space? Well, [here's](https://blogs.msdn.microsoft.com/ntdebugging/2007/01/04/desktop-heap-overview/) a good overview. 
- The instructions are actually in the blog post, but generally the workaround if you want to run that dense is to increase the heap size of either the desktop heap (second section) or the noninteractive heap (third value).  You should be able to try this out locally to repro the issue before you go considering it for prod :)

```
HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control\Session Manager\SubSystems\Windows
```
There's a fragment that looks something like this:  "SharedSection=1024,20480,768"
You can set the third value to something like 4096 and reboot the machine.

- But doing that in Azure repeatably via something like a VMSS Extension is hard, so there's another option, which is to change how ServiceFabric launches your process. Most processes don't actually need a console window, so you can launch them in a different session. 
- To do _that_ you can change (upgrade or redeploy, kinda your call) the EnableActivateNoWindow setting in your cluster manifest to True. (instructions on [this page](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-fabric-settings#section-name-hosting), link is to the specific section).
- This usually gets you to a place where you're not consuming as much space, and things will run fine. If you still find yourself running out of heap though you'll have to go back and increase the heap size.
- Side effect of activating without a console window that is that you won't get Ctrl+C signal in your host process during process shutdown anymore. Most services don't rely on this, but it's an FYI in case you are. 

Final thing: Designing for scalability
 - Generally we try to recommend a more adaptive model for creating services, rather than creating them all up front. This might help you out so that you don't have to run as many services until your load actually shows up. [This doc](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-concepts-scalability) has a good overview of some other patterns you might consider to see if they can work for you.


Hope all this helps!

**Comment by masnider** [comment]
@francescocristallo can you let us know if my (long) answer helped you? Did you figure out what's up and are you good now?

**Comment by francescocristallo** [comment]
Hi and thanks for your time in replying me.

I did many tests and debugs, I think the main problem was the RAM of the machines.
Using more RAM I was able to use more partitions. I didn't suspect any single working set would use 60MB of RAM. 
The reason I need all these partitions is for the concurrent queues, I need to process 10k queues per second and I wanted to scale it creating more partitions. Any single queue needs around 800ms to process, so I either create multiple clusters and use Traffic manager to send the traffic to them or I use more partitions in a single Cluster.

I think I'll test using 100 partitions with more threads in any single RynAsync and see if it works.

Thanks

**Comment by justinSelf** [comment]
@masnider 

> You'd need to type "NodeType == Workers"

Shouldn't it be `NodeTypeName`?

**Comment by masnider** [comment]
@justinSelf In Azure the clusters get automatically set up with NodeTypeName == "whatever" but this is a leftover unnecessary thing and doesn't happen in all environments. NodeType is a built in and automatic node property that works everywhere and is what we document and recommend. 

@francescocristallo I'll close this down for now since you've probably gotten to the root of the issue (lack of memory or heap). Feel free to reopen if you're still seeing issues or to create another issue if you have further problems.
